File: scripts/governance/check_security_critical.py
#!/usr/bin/env python3
"""Security Review Gate — @security-critical blocks require human review"""

import subprocess
import sys
from pathlib import Path


def get_staged_files() -> list[Path]:
    try:
        result = subprocess.run(
            ["git", "diff", "--cached", "--name-only", "--diff-filter=ACMR"],
            capture_output=True,
            text=True,
        )
        return [
            Path(f) for f in result.stdout.strip().split("\n") if f and Path(f).exists()
        ]
    except Exception:
        return []


files = get_staged_files()
violations = []

SOURCE_EXTS = {".py", ".ts", ".js", ".tsx", ".jsx"}
for filepath in files:
    if filepath.suffix not in SOURCE_EXTS:
        continue
    content = filepath.read_text()
    if "@security-critical" in content and "security-review-approved:" not in content:
        violations.append(
            f"  {filepath}: has @security-critical but no security-review-approved comment"
        )

if violations:
    print("✗ Security Review Gate:", file=sys.stderr)
    for v in violations:
        print(v, file=sys.stderr)
    print("  Add: # security-review-approved: <reviewer> <date>", file=sys.stderr)
    sys.exit(1)

print("✓ Security review: no unapproved security-critical blocks")

File: scripts/governance/check-security-critical.js
#!/usr/bin/env node
/**
 * Security Review Gate — @security-critical blocks require human review comment.
 * Staged files with @security-critical annotation must have a reviewer override:
 * // security-review-approved: <reviewer> <date>
 */

const fs = require('fs');
const { execSync } = require('child_process');

function getStagedFiles() {
  try {
    return execSync('git diff --cached --name-only --diff-filter=ACMR', { encoding: 'utf8' })
      .trim().split('\n').filter(Boolean);
  } catch { return []; }
}

// Only check source code files, not config/yaml/markdown
const SOURCE_EXTS = ['.ts', '.tsx', '.js', '.jsx', '.py'];
const files = getStagedFiles()
  .filter(f => fs.existsSync(f) && SOURCE_EXTS.some(ext => f.endsWith(ext)));
const violations = [];

for (const file of files) {
  const content = fs.readFileSync(file, 'utf8');
  if (content.includes('@security-critical')) {
    if (!content.includes('security-review-approved:')) {
      violations.push(`  ${file}: has @security-critical but no security-review-approved comment`);
    }
  }
}

if (violations.length > 0) {
  console.error('✗ Security Review Gate:');
  violations.forEach(v => console.error(v));
  console.error('  Add: // security-review-approved: <reviewer> <date>');
  process.exit(1);
}

console.log('✓ Security review: no unapproved security-critical blocks');
process.exit(0);
